src/memprof_tracer.ml

```diff
@@ -104,68 +94,64 @@ let start ?(report_exn=default_report_exn) ~sampling_rate trace =
         match Trace.Writer.put_collect trace (Trace.Timestamp.now ()) id with
         | () -> unlock_tracer s
         | exception e -> mark_failed s e) } in
-  curr_active_tracer := Some s;
-  bytes_before_ext_sample := draw_sampler_bytes s;
-  Gc.Memprof.start
-    ~sampling_rate
-    ~callstack_size:max_int
-    tracker;
+  Atomic.set curr_active_tracer (Some s);
+  Atomic.set bytes_before_ext_sample (draw_sampler_bytes s);
+  ignore (Gc.Memprof.start ~sampling_rate ~callstack_size:max_int tracker);
   s
 
 let stop s =
   if not s.stopped then begin
     s.stopped <- true;
     Gc.Memprof.stop ();
-    if lock_tracer s then begin
-      try Trace.Writer.close s.trace with e -> mark_failed s e
-    end;
-    curr_active_tracer := None
+    Mutex.protect s.mutex (fun () ->
+        try Trace.Writer.close s.trace
+        with e ->
+          (s.failed <- true; s.report_exn e);
+        Atomic.set curr_active_tracer None
+      )
   end
 
 let[@inline never] ext_alloc_slowpath ~bytes =
-  match !curr_active_tracer with
-  | None -> bytes_before_ext_sample := max_int; None
+  match Atomic.get curr_active_tracer with
+  | None -> Atomic.set bytes_before_ext_sample max_int; None
   | Some s ->
-    if lock_tracer_ext s then begin
+    if lock_tracer s then begin
       match
         let bytes_per_word = Sys.word_size / 8 in
         (* round up to an integer number of words *)
         let size_words = (bytes + bytes_per_word - 1) / bytes_per_word in
-        let samples = ref 0 in
-        while !bytes_before_ext_sample <= 0 do
-          bytes_before_ext_sample :=
-            !bytes_before_ext_sample + draw_sampler_bytes s;
-          incr samples
+        let samples = Atomic.make 0 in
+        while Atomic.get bytes_before_ext_sample <= 0 do
+          ignore (Atomic.fetch_and_add bytes_before_ext_sample (draw_sampler_bytes s));
+          Atomic.incr samples
         done;
-        assert (!samples > 0);
+        assert (Atomic.get samples > 0);
         let callstack = Printexc.get_callstack max_int in
         Some (Trace.Writer.put_alloc_with_raw_backtrace s.trace
                 (Trace.Timestamp.now ())
                 ~length:size_words
-                ~nsamples:!samples
+                ~nsamples:(Atomic.get samples)
```

Comment:
Seems that `samples` can in fact be a `ref` as it was originally, because it's only used locally.